File: codes/python/99.恢复二叉搜索树.py


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
	def recoverTree(self, root: Optional[TreeNode]) -> None:
		"""
		Do not return anything, modify root in-place instead.
		"""
		if not root:
			return
		nums = []
		self.inorder(root, nums)
		x, y = self.find_two_swapped(nums)
		self.recover(root, x, y)
	
	def recover(self, root: Optional[TreeNode], x: int, y: int) -> None:
		if not root:
			return
		if root.val == x:
			root.val = y
		elif root.val == y:
			root.val = x
		self.recover(root.left, x, y)
		self.recover(root.right, x, y)

	def inorder(self, root: Optional[TreeNode], nums: List[int]) -> None:
		if not root:
			return
		self.inorder(root.left, nums)
		nums.append(root.val)
		self.inorder(root.right, nums)

	def find_two_swapped(self, nums: List[int]) -> (int, int):
		x = y = -1
		for i in range(len(nums) - 1):
			if nums[i + 1] < nums[i]:
				y = i+1
				if x == -1:
					x = i
				else:
					break
		return nums[x], nums[y]
	

File: codes/java/99.恢复二叉搜索树.java

import java.util.Arrays;
import java.util.List;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
	public void recoverTree(TreeNode root) {
		// 首先，我们利用中序遍历，得到一个递增的数组
		List<Integer> nums = new ArrayList<>();
		inorder(root, nums);
		// 接着，我们找到两个被交换的元素
		int[] swapped = findTwoSwapped(nums);
		// 最后，我们交换这两个元素
		recover(root, swapped[0], swapped[1]);
	}

	public void inorder(TreeNode root, List<Integer> nums) {
		if (root == null) {
			return;
		}
		inorder(root.left, nums);
		nums.add(root.val);
		inorder(root.right, nums);
	}

	public int[] findTwoSwapped(List<Integer> nums) {
		int n = nums.size();
		int index1 = -1, index2 = -1;
		for (int i = 0; i < n - 1; ++i) {
			if (nums.get(i + 1) < nums.get(i)) {
				index2 = i + 1;
				// first swap occurence
				if (index1 == -1) {
					index1 = i;
				} else {
					// second swap occurence
					break;
				}
			}
		}
		return new int[] { nums.get(index1), nums.get(index2) };
	}

	public void recover(TreeNode r, int x, int y) {
		if (r == null) {
			return;
		}
		if (r.val == x || r.val == y) {
			r.val = r.val == x ? y : x;
		}
		recover(r.left, x, y);
		recover(r.right, x, y);
	}
}